zaza/openstack/utilities/parallel_series_upgrade.py

```diff
@@ -0,0 +1,639 @@
+    :returns: Nothing
+    :trype: None
+    """
+    subordinate_pauses = []
+    leader_pauses = []
+    for unit in units:
+        if pause_non_leader_subordinate:
+            if status["units"][unit].get("subordinates"):
+                for subordinate in status["units"][unit]["subordinates"]:
+                    _app = subordinate.split('/')[0]
+                    if _app in SUBORDINATE_PAUSE_RESUME_BLACKLIST:
+                        logging.info("Skipping pausing {} - blacklisted"
+                                     .format(subordinate))
+                    else:
+                        logging.info("Pausing {}".format(subordinate))
+                        subordinate_pauses.append(model.async_run_action(
+                            subordinate, "pause", action_params={}))
+        if pause_non_leader_primary:
+            logging.info("Pausing {}".format(unit))
+            leader_pauses.append(
+                model.async_run_action(unit, "pause", action_params={}))
+    await asyncio.gather(*leader_pauses)
+    await asyncio.gather(*subordinate_pauses)
+
+
+def get_leader_and_non_leaders(status):
+    """Get the leader and non-leader Juju units.
+
+    This function returns a tuple that looks like:
+
+    ({
+        'unit/1': juju.Unit,
+    },
+    {
+        'unit/0': juju.Unit,
+        'unit/2': juju.unit,
+    })
+
+    The first entry of this tuple is the leader, and the second is
+    all non-leader units.
+
+    :returns: A tuple of dicts identifying leader and non-leaders
+    :rtype: Dict[str, List[juju.Unit]]
+    """
+    leader = None
+    non_leaders = {}
+    for name, unit in status["units"].items():
+        if unit.get("leader"):
+            leader = {name: unit}
+        else:
+            non_leaders[name] = unit
+    return (leader, non_leaders)
+
+
+async def prepare_series_upgrade(machine, to_series):
+    """Execute juju series-upgrade prepare on machine.
+
+    NOTE: This is a new feature in juju behind a feature flag and not yet in
+    libjuju.
+    export JUJU_DEV_FEATURE_FLAGS=upgrade-series
+    :param machine_num: Machine number
+    :type machine_num: str
+    :param to_series: The series to which to upgrade
+    :type to_series: str
+    :returns: None
+    :rtype: None
+    """
+    logging.info("Preparing series upgrade for: {}".format(machine))
+    await series_upgrade_utils.async_prepare_series_upgrade(
+        machine, to_series=to_series)
+
+
+async def reboot(machine):
+    """Reboot the named machine.
+
+    :param machine: Machine to reboot
+    :type machine: str
+    :returns: Nothing
+    :rtype: None
+    """
+    try:
+        await run_on_machine(machine, 'sudo init 6 & exit')
+        # await run_on_machine(unit, "sudo reboot && exit")
+    except subprocess.CalledProcessError as e:
+        logging.warn("Error doing reboot: {}".format(e))
+        pass
+
+
+async def run_on_machine(machine, command, model_name=None, timeout=None):
```

Comment:
Is there not a method in zaza.model that does this already?

Reply:
Currently, nope; there's just `run_on_unit` and `run_on_leader`, both of which expect an actual application unit name